codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -184,15 +183,25 @@ public void goAwaySent(int lastKnownStream, long errorCode, ByteBuf debugData) {
         }
     }
 
-    private void removeStream(DefaultStream stream) {
-        // Notify the listeners of the event first.
-        for (Listener listener : listeners) {
-            listener.onStreamRemoved(stream);
-        }
+    /**
+     * Closed streams may stay in the priority tree if they have dependents that are in prioritizable states.
+     * When a stream is requested to be removed we can only actually remove that stream when there are no more
+     * prioritizable children.
+     * (see [1] {@link Http2Stream#prioritizableForTree()} and [2] {@link DefaultStream#removeChild(DefaultStream)}).
+     * When a priority tree edge changes we also have to re-evaluate viable nodes
+     * (see [3] {@link DefaultStream#takeChild(DefaultStream, boolean, List)}).
+     * @param stream The stream to remove.
+     */
+    void removeStream(DefaultStream stream) {
+        // [1] Check if this stream can be removed because it has no prioritizable descendants.
+        if (stream.parent().removeChild(stream)) {
+            // Remove it from the map and priority tree.
+            streamMap.remove(stream.id());
 
-        // Remove it from the map and priority tree.
```

Comment:
@nmittler - Note that I change the notification for `onStreamRemoved` to be after it is removed from the priority tree and stream map.  I think this lines up with how the other listeners are used (notify after event takes place) and also the comments on the `onStreamRemoved` method.